src/common/snippets/src/pass/canonicalization.cpp

```diff
@@ -0,0 +1,86 @@
+    }
+}
+
+bool pass::Canonicalization::run_on_model(const std::shared_ptr<ov::Model>& m) {
+    RUN_ON_MODEL_SCOPE(Canonicalization);
+    OV_ITT_SCOPED_TASK(ov::pass::itt::domains::SnippetsTransform, "Snippets::Canonicalization")
+    bool is_modified = false;
+    const ParameterVector& params = m->get_parameters();
+    OPENVINO_ASSERT(m_in_shapes.size() == params.size(),
+                    "Number of parameters for snippet doesn't match passed to the Canonicalization pass. ",
+                    "Expected: ", m_in_shapes.size(), " Got: ", params.size(), ".");
+
+    // Note that shape rank also incorporates layout, so NCHW16c would have shape rank 5
+    auto is_blocked_layout = [](const Layout& l) {
+        return l.size() != std::set<size_t>(l.begin(), l.end()).size();
+    };
+    auto get_rank = [](const Layout& l, const Layout& r) {
+        return l.size() < r.size();
+    };
+    // Layout with the max rank
+    const auto& max_rank_it = std::max_element(m_in_layouts.begin(), m_in_layouts.end(), get_rank);
+    Layout base_layout = *max_rank_it;
+    size_t max_rank = base_layout.size();
+    const bool base_is_blocked = is_blocked_layout(base_layout);
+
+    for (size_t i = 0; i < m_in_layouts.size(); i++) {
+        const auto& i_layout = m_in_layouts[i];
+        const auto& i_shape = m_in_shapes[i];
+        const auto i_rank = i_layout.size();
+        const bool i_is_blocked = is_blocked_layout(i_layout);
+        // Canonicalization logic briefly:
+        // * If this input is blocked => Reshape corresponding input parameter, so the following transformations
+        //   will work with a shape of a larger rank. In dynamic case, this shape will be updated during shapeInfer()
+        //   call, but the important thing is that the shape rank won't change.
+        // * If some of the input shapes is blocked (=> base_is_blocked), but this input is planar,
+        //   then insert RankNormalization op after this input. This is needed, so all shapes inside the body have
+        //   similar ranks.
+        if (i_is_blocked) {
+            OPENVINO_ASSERT(base_is_blocked && i_rank == max_rank, "If this shape is blocked, base must also be blocked");
+            params[i]->set_partial_shape(snippets::utils::vdims_to_pshape(i_shape));
+            is_modified = true;
+        } else if (i_rank < max_rank) {
+            size_t num_append = base_is_blocked;
+            OPENVINO_ASSERT(max_rank >= i_rank + num_append, "Unsupported blocked shapes combination in canonicalization");
+            size_t num_prepend = max_rank - i_rank - num_append;
+            // here I should insert unsqueeze operation after the corresponding input
+            for (const auto& out : params[i]->outputs()) {
+                const auto& target_inputs = out.get_target_inputs();
+                auto rank_norm = std::make_shared<op::RankNormalization>(out, num_prepend, num_append);
```

Comment:
I also don't like this op, but it's the most convenient way to avoid canonicalization for every new shape. On the bright side, this op allows us to support mixed layouts
We can't remove it from the LIR, because shape inference is performed on LIR, and we need it for correct shape infer. Moreover, we even can't remove it before shape-sensitive control flow pipeline for the same reason (there will be incorrect ranks and broadcasting rules). The only alternative option is to embed semantics of rank normalization into `Parameter`, but this wouldn't work if a `Subgraph`'s subgraph is performed on planar shapes (and the rest is on bloked ones). Broadcasting is implemented as a stand-alone op for the same reasons btw

I agree that this looks ugly at first, but the more you think about alternatives, the prettier it gets 😁 . Anyway I'm open for new ideas, so let's discuss your suggestions offline